collect_app/src/main/java/org/odk/collect/android/utilities/QRCodeUtils.java

```diff
@@ -91,50 +84,88 @@ private static BinaryBitmap getBinaryBitmap(Bitmap bitmap) {
         return new BinaryBitmap(new HybridBinarizer(source));
     }
 
-    public static File saveBitmapToCache(Bitmap qrCode) throws IOException {
-        //Save the bitmap to a file
-        File cache = Collect.getInstance().getApplicationContext().getExternalCacheDir();
-        File shareFile = new File(cache, "shareImage.jpeg");
-        FileOutputStream out = new FileOutputStream(shareFile);
-        qrCode.compress(Bitmap.CompressFormat.JPEG, 100, out);
-        out.close();
-        return shareFile;
+    public static Bitmap generateQRBitMap(String data, int sideLength) throws IOException, WriterException {
+        final long time = System.currentTimeMillis();
+        String compressedData = CompressionUtils.compress(data);
+
+        // Maximum capacity for QR Codes is 4,296 characters (Alphanumeric)
+        if (compressedData.length() > 4000) {
+            throw new IOException(Collect.getInstance().getString(R.string.encoding_max_limit));
+        }
+
+        Map<EncodeHintType, ErrorCorrectionLevel> hints = new HashMap<>();
+        hints.put(EncodeHintType.ERROR_CORRECTION, ErrorCorrectionLevel.L);
+        QRCodeWriter qrCodeWriter = new QRCodeWriter();
+        BitMatrix bitMatrix = qrCodeWriter.encode(compressedData, BarcodeFormat.QR_CODE, sideLength, sideLength, hints);
```

Comment:
This takes on the order of ten seconds on my Nexus 7. Are we doing whatever we can to get good performance?

Reply:
1. We are using the lowest  `ErrorCorrectionLevel` while creating the QRBitmap.
2. Preferences are compressed using deflate to reduce the size of content

On the devices I tested, it took a few hundred milliseconds for encoding default settings to QR Code

Reply:
`Nexus 5` :  QR Code generation took : 2453 ms

Reply:
OK. Thanks for checking.